File: src/ee/admob/private/AdMob.swift
//
//  AdMob.swift
//  ee-x-366f64ba
//
//  Created by eps on 6/24/20.
//

import Foundation
import GoogleMobileAds

private let kPrefix = "AdMob"
private let kInitialize = "\(kPrefix)Initialize"
private let kGetEmulatorTestDeviceHash = "\(kPrefix)GetEmulatorTestDeviceHash"
private let kAddTestDevice = "\(kPrefix)AddTestDevice"
private let kGetBannerAdSize = "\(kPrefix)GetBannerAdSize"
private let kCreateBannerAd = "\(kPrefix)CreateBannerAd"
private let kDestroyBannerAd = "\(kPrefix)DestroyBannerAd"
private let kCreateNativeAd = "\(kPrefix)CreateNativeAd"
private let kDestroyNativeAd = "\(kPrefix)DestroyNativeAd"
private let kCreateInterstitialAd = "\(kPrefix)CreateInterstitialAd"
private let kDestroyInterstitialAd = "\(kPrefix)DestroyInterstitialAd"
private let kCreateRewardedAd = "\(kPrefix)CreateRewardedAd"
private let kDestroyRewardedAd = "\(kPrefix)DestroyRewardedAd"

@objc(EEAdMob)
class AdMob: NSObject, IPlugin {
    private let _bridge: IMessageBridge
    private let _bannerHelper = AdMobBannerHelper()
    private var _testDevices: [String] = []
    private var _bannerAds: [String: AdMobBannerAd] = [:]
    private var _nativeAds: [String: AdMobNativeAd] = [:]
    private var _interstitialAds: [String: AdMobInterstitialAd] = [:]
    private var _rewardedAds: [String: AdMobRewardedAd] = [:]

    required init(_ bridge: IMessageBridge) {
        _bridge = bridge
        super.init()
        registerHandlers()
    }

    func destroy() {
        deregisterHandlers()
        _bannerAds.values.forEach { $0.destroy() }
        _bannerAds.removeAll()
        _nativeAds.values.forEach { $0.destroy() }
        _nativeAds.removeAll()
        _interstitialAds.values.forEach { $0.destroy() }
        _interstitialAds.removeAll()
        _rewardedAds.values.forEach { $0.destroy() }
        _rewardedAds.removeAll()
    }

    func registerHandlers() {
        _bridge.registerHandler(kInitialize) { _ in
            self.initialize()
            return ""
        }
        _bridge.registerHandler(kGetEmulatorTestDeviceHash) { _ in
            self.emulatorTestDeviceHash
        }
        _bridge.registerHandler(kAddTestDevice) { message in
            self.addTestDevice(message)
            return ""
        }
        _bridge.registerHandler(kGetBannerAdSize) { message in
            let sizeId = Int(message) ?? -1
            let size = self.getBannerAdSize(sizeId)
            return EEJsonUtils.convertDictionary(toString: [
                "width": size.width,
                "height": size.height
            ])
        }
        _bridge.registerHandler(kCreateBannerAd) { message in
            let dict = EEJsonUtils.convertString(toDictionary: message)
            guard let adId = dict["ad_id"] as? String else {
                assert(false, "Invalid argument")
            }
            guard let sizeId = dict["ad_size"] as? Int else {
                assert(false, "Invalid argument")
            }
            let adSize = self._bannerHelper.getAdSize(sizeId)
            return Utils.toString(self.createBannerAd(adId, adSize))
        }
        _bridge.registerHandler(kDestroyBannerAd) { message in
            Utils.toString(self.destroyBannerAd(message))
        }
        _bridge.registerHandler(kCreateNativeAd) { message in
            let dict = EEJsonUtils.convertString(toDictionary: message)
            guard let adId = dict["ad_id"] as? String else {
                assert(false, "Invalid argument")
            }
            guard let layoutName = dict["layout_name"] as? String else {
                assert(false, "Invalid argument")
            }
            return Utils.toString(self.createNativeAd(adId, layoutName))
        }
        _bridge.registerHandler(kDestroyNativeAd) { message in
            Utils.toString(self.destroyNativeAd(message))
        }
        _bridge.registerHandler(kCreateInterstitialAd) { message in
            Utils.toString(self.createInterstitialAd(message))
        }
        _bridge.registerHandler(kDestroyInterstitialAd) { message in
            Utils.toString(self.destroyInterstitialAd(message))
        }
        _bridge.registerHandler(kCreateRewardedAd) { message in
            Utils.toString(self.createRewardedAd(message))
        }
        _bridge.registerHandler(kDestroyRewardedAd) { message in
            Utils.toString(self.destroyRewardedAd(message))
        }
    }

    func deregisterHandlers() {
        _bridge.deregisterHandler(kInitialize)
        _bridge.deregisterHandler(kGetEmulatorTestDeviceHash)
        _bridge.deregisterHandler(kAddTestDevice)
        _bridge.deregisterHandler(kGetBannerAdSize)
        _bridge.deregisterHandler(kCreateBannerAd)
        _bridge.deregisterHandler(kDestroyBannerAd)
        _bridge.deregisterHandler(kCreateNativeAd)
        _bridge.deregisterHandler(kDestroyNativeAd)
        _bridge.deregisterHandler(kCreateInterstitialAd)
        _bridge.deregisterHandler(kDestroyInterstitialAd)
        _bridge.deregisterHandler(kCreateRewardedAd)
        _bridge.deregisterHandler(kDestroyRewardedAd)
    }

    func initialize() {
        Thread.runOnMainThread {
            GADMobileAds.sharedInstance().start(completionHandler: nil)
        }
    }

    var emulatorTestDeviceHash: String {
        return kGADSimulatorID as? String ?? ""
    }

    func addTestDevice(_ hash: String) {
        Thread.runOnMainThread {
            self._testDevices.append(hash)
            GADMobileAds.sharedInstance().requestConfiguration.testDeviceIdentifiers = self._testDevices
        }
    }

    func getBannerAdSize(_ sizeId: Int) -> CGSize {
        return _bannerHelper.getSize(index: sizeId)
    }

    func createBannerAd(_ adId: String, _ size: GADAdSize) -> Bool {
        if _bannerAds.contains(where: { key, _ in key == adId }) {
            return false
        }
        let ad = AdMobBannerAd(_bridge, adId, size)
        _bannerAds[adId] = ad
        return true
    }

    func destroyBannerAd(_ adId: String) -> Bool {
        guard let ad = _bannerAds[adId] else {
            return false
        }
        ad.destroy()
        _bannerAds.removeValue(forKey: adId)
        return true
    }

    func createNativeAd(_ adId: String, _ layoutName: String) -> Bool {
        if _nativeAds.contains(where: { key, _ in key == adId }) {
            return false
        }
        let ad = AdMobNativeAd(_bridge, adId, layoutName)
        _nativeAds[adId] = ad
        return true
    }

    func destroyNativeAd(_ adId: String) -> Bool {
        guard let ad = _nativeAds[adId] else {
            return false
        }
        ad.destroy()
        _nativeAds.removeValue(forKey: adId)
        return true
    }

    func createInterstitialAd(_ adId: String) -> Bool {
        if _interstitialAds.contains(where: { key, _ in key == adId }) {
            return false
        }
        let ad = AdMobInterstitialAd(_bridge, adId)
        _interstitialAds[adId] = ad
        return true
    }

    func destroyInterstitialAd(_ adId: String) -> Bool {
        guard let ad = _interstitialAds[adId] else {
            return false
        }
        ad.destroy()
        _interstitialAds.removeValue(forKey: adId)
        return true
    }

    func createRewardedAd(_ adId: String) -> Bool {
        if _rewardedAds.contains(where: { key, _ in key == adId }) {
            return false
        }
        let ad = AdMobRewardedAd(_bridge, adId)
        _rewardedAds[adId] = ad
        return true
    }

    func destroyRewardedAd(_ adId: String) -> Bool {
        guard let ad = _rewardedAds[adId] else {
            return false
        }
        ad.destroy()
        _rewardedAds.removeValue(forKey: adId)
        return true
    }
}

File: src/ts/src/ad_mob/internal/AdMob.ts
import {
    FullScreenAdResult,
    IAd,
    IBannerAd,
    IFullScreenAd,
} from "../../ads";
import {
    DefaultBannerAd,
    DefaultFullScreenAd,
    GuardedBannerAd,
    GuardedFullScreenAd,
    IAsyncHelper,
    MediationManager,
} from "../../ads/internal";
import {
    ILogger,
    IMessageBridge,
    Utils,
} from "../../core";
import { AdMobBannerAdSize } from "../AdMobBannerAdSIze";
import { IAdMob } from "../IAdMob";

type Destroyer = () => void;

export class AdMob implements IAdMob {
    private readonly kTag = `AdMob`;
    private readonly kPrefix = "AdMobBridge";
    private readonly kInitialize = this.kPrefix + "Initialize";
    private readonly kGetEmulatorTestDeviceHash = this.kPrefix + "GetEmulatorTestDeviceHash";
    private readonly kAddTestDevice = this.kPrefix + "AddTestDevice";
    private readonly kGetBannerAdSize = this.kPrefix + "GetBannerAdSize";
    private readonly kCreateBannerAd = this.kPrefix + "CreateBannerAd";
    private readonly kDestroyBannerAd = this.kPrefix + "DestroyBannerAd";
    private readonly kCreateNativeAd = this.kPrefix + "CreateNativeAd";
    private readonly kDestroyNativeAd = this.kPrefix + "DestroyNativeAd";
    private readonly kCreateAppOpenAd = this.kPrefix + "CreateAppOpenAd";
    private readonly kDestroyAppOpenAd = this.kPrefix + "DestroyAppOpenAd";
    private readonly kCreateInterstitialAd = this.kPrefix + "CreateInterstitialAd";
    private readonly kDestroyInterstitialAd = this.kPrefix + "DestroyInterstitialAd";
    private readonly kCreateRewardedAd = this.kPrefix + "CreateRewardedAd";
    private readonly kDestroyRewardedAd = this.kPrefix + "DestroyRewardedAd";

    private readonly _bridge: IMessageBridge;
    private readonly _logger: ILogger;
    private readonly _destroyer: Destroyer;
    private readonly _ads: { [index: string]: IAd };
    private readonly _displayer: IAsyncHelper<FullScreenAdResult>;

    public constructor(bridge: IMessageBridge, logger: ILogger, destroyer: Destroyer) {
        this._bridge = bridge;
        this._logger = logger;
        this._destroyer = destroyer;
        this._logger.debug(`${this.kTag}: constructor`);
        this._ads = {};
        this._displayer = MediationManager.getInstance().adDisplayer;
    }

    public destroy(): void {
        this._logger.debug(`${this.kTag}: destroy`);
        for (const id in this._ads) {
            const ad = this._ads[id];
            ad.destroy();
            delete this._ads[id];
        }
        this._destroyer();
    }

    public async initialize(): Promise<boolean> {
        const response = await this._bridge.callAsync(this.kInitialize);
        return Utils.toBool(response);
    }

    public getEmulatorTestDeviceHash(): string {
        return this._bridge.call(this.kGetEmulatorTestDeviceHash);
    }

    public addTestDevice(hash: string): void {
        this._bridge.call(this.kAddTestDevice, hash);
    }

    private getBannerAdSize(adSize: AdMobBannerAdSize): [number, number] {
        const response = this._bridge.call(this.kGetBannerAdSize, `${adSize}`);
        const json: {
            width: number,
            height: number,
        } = JSON.parse(response);
        return [json.width, json.height];
    }

    public createBannerAd(adId: string, adSize: AdMobBannerAdSize): IBannerAd {
        this._logger.debug(`${this.kTag}: createBannerAd: id = ${adId} size = ${adSize}`);
        if (this._ads[adId]) {
            return this._ads[adId] as IBannerAd;
        }
        const request = {
            [`adId`]: adId,
            [`adSize`]: parseInt(`${adSize}`),
        };
        const response = this._bridge.call(this.kCreateBannerAd, JSON.stringify(request));
        if (!Utils.toBool(response)) {
            throw new Error(`Failed to create banner ad: ${adId}`);
        }
        const size = this.getBannerAdSize(adSize);
        const ad = new GuardedBannerAd(
            new DefaultBannerAd("AdMobBannerAd", this._bridge, this._logger,
                () => this.destroyAd(this.kDestroyBannerAd, adId),
                adId, size));
        this._ads[adId] = ad;
        return ad;
    }

    public createAppOpenAd(adId: string): IFullScreenAd {
        return this.createFullScreenAd(this.kCreateAppOpenAd, adId, () =>
            new DefaultFullScreenAd("AdMobAppOpenAd", this._bridge, this._logger, this._displayer,
                () => this.destroyAd(this.kDestroyAppOpenAd, adId),
                _ => FullScreenAdResult.Completed,
                adId))
    }

    public createInterstitialAd(adId: string): IFullScreenAd {
        return this.createFullScreenAd(this.kCreateInterstitialAd, adId, () =>
            new DefaultFullScreenAd("AdMobInterstitialAd", this._bridge, this._logger, this._displayer,
                () => this.destroyAd(this.kDestroyInterstitialAd, adId),
                _ => FullScreenAdResult.Completed,
                adId))
    }

    public createRewardedAd(adId: string): IFullScreenAd {
        return this.createFullScreenAd(this.kCreateRewardedAd, adId, () =>
            new DefaultFullScreenAd("AdMobRewardedAd", this._bridge, this._logger, this._displayer,
                () => this.destroyAd(this.kDestroyRewardedAd, adId),
                message => Utils.toBool(message)
                    ? FullScreenAdResult.Completed
                    : FullScreenAdResult.Canceled,
                adId))
    }

    private createFullScreenAd(handlerId: string, adId: string, creator: () => IFullScreenAd): IFullScreenAd {
        this._logger.debug(`${this.kTag}: createFullScreenAd: id = ${adId}`);
        if (this._ads[adId]) {
            return this._ads[adId] as IFullScreenAd;
        }
        const response = this._bridge.call(handlerId, adId);
        if (!Utils.toBool(response)) {
            throw new Error(`Failed to create fullscreen ad: ${adId}`);
        }
        const ad = new GuardedFullScreenAd(creator());
        this._ads[adId] = ad;
        return ad;
    }

    private destroyAd(handlerId: string, adId: string): boolean {
        this._logger.debug(`${this.kTag}: destroyAd: id = ${adId}`);
        if (!this._ads[adId]) {
            return false;
        }
        const response = this._bridge.call(handlerId, adId);
        if (!Utils.toBool(response)) {
            // Assert.
            return false;
        }
        delete this._ads[adId];
        return true;
    }
}